File: glmm/src/FileIo/MatrixReader.java
package FileIo;
import java.io.BufferedReader;
import java.util.ArrayList;
public class MatrixReader
{
	private String fileName;
	private double X[][];
	
	public MatrixReader(String fName)
	{
		fileName = fName;
		X = null;
	}//End of this method.
	
	public double[][] getData()
	{
		return X;
	}
	/*
	 * This method is used to write a Matrix into a specified file.
	 */
	public void read()
	{
		String strSplitter = ",";
		BufferedReader fin 	= FileIOHelper.openFileForRead(fileName);
		String sentence 	= null;
		String tokens[] 	= null;
		ArrayList<double[]> dataM = new ArrayList<double[]>();
		// Read data from the input file.
		while ((sentence = FileIOHelper.readLine(fin)) != null) 
		{
			tokens = sentence.trim().split(strSplitter);
			int n = tokens.length;
			double rowM[] = new double[n];
			for(int i = 0; i < n; i++)
				rowM[i] = Double.valueOf(tokens[i]);
			dataM.add(rowM);
		}
		FileIOHelper.closeFile(fin);
		int n = dataM.size();
		int m = dataM.get(0).length;
		X = new double[n][m];
		for(int i = 0; i < n; i++)
			X[i] = dataM.get(i);
	}//End of this method.
}//End of this class.

File: GLMM_C/src/MatrixReader.cpp
#include "MatrixReader.h"

using namespace std;
MatrixReader::MatrixReader(std::string in){
	_fileName=in;
}

const std::vector < std::vector<double> > & MatrixReader::getData()const {
	return _X;
}
void MatrixReader::read(){
	std::string strSplitter=",";
	std::unique_ptr<std::ifstream> fin=FileIOHelper::openFileForRead(_fileName);
	std::string sentence;
	std::vector< std::vector<double> > dataM;
	
	//initial row vector of doubles
	//preread
	sentence=FileIOHelper::readLine(fin);
	sentence=nstring::trim(sentence,",");
	sentence=nstring::trim(sentence);
	while(sentence.length()!=0){ //rows
		std::vector<double> tokensDoubles;
		std::stringstream ss(nstring::trim(sentence));
		if(ss){
			std::string token;
            while(std::getline(ss,token,',')){ //columns
				std::istringstream attrSS(token);
				double attrD;
				attrSS >> attrD;
				tokensDoubles.push_back(attrD);
            }
		}
		if(tokensDoubles.size()!=0)
			_X.push_back(tokensDoubles);//push row
		sentence=nstring::trim(FileIOHelper::readLine(fin),",");
		sentence=nstring::trim(sentence);
	}

	FileIOHelper::closeFile(fin);
}


